src/Development/IDE/Core/Completions.hs

```diff
@@ -0,0 +1,565 @@
+            -- like '-> '
+            d = T.length fullLine - T.length (stripTypeStuff partialLine)
+        in Position l (c - d)
+
+      filtModNameCompls =
+        map mkModCompl
+          $ mapMaybe (T.stripPrefix enteredQual)
+          $ Fuzzy.simpleFilter fullPrefix allModNamesAsNS
+
+      filtCompls = map Fuzzy.original $ Fuzzy.filter prefixText ctxCompls "" "" label False
+        where
+          isTypeCompl = isTcOcc . occName . origName
+          -- completions specific to the current context
+          ctxCompls' = case context of
+                        TypeContext -> filter isTypeCompl compls
+                        ValueContext -> filter (not . isTypeCompl) compls
+                        _ -> filter (not . isTypeCompl) compls
+          -- Add whether the text to insert has backticks
+          ctxCompls = map (\comp -> comp { isInfix = infixCompls }) ctxCompls'
+
+          infixCompls :: Maybe Backtick
+          infixCompls = isUsedAsInfix fullLine prefixModule prefixText (VFS.cursorPos prefixInfo)
+
+          compls = if T.null prefixModule
+            then unqualCompls
+            else Map.findWithDefault [] prefixModule $ getQualCompls qualCompls
+
+      filtListWith f list =
+        [ f label
+        | label <- Fuzzy.simpleFilter fullPrefix list
+        , enteredQual `T.isPrefixOf` label
+        ]
+
+      filtListWithSnippet f list suffix =
+        [ toggleSnippets caps withSnippets (f label (snippet <> suffix))
+        | (snippet, label) <- list
+        , Fuzzy.test fullPrefix label
+        ]
+
+      filtImportCompls = filtListWith (mkImportCompl enteredQual) importableModules
+      filtPragmaCompls = filtListWithSnippet mkPragmaCompl validPragmas
+      filtOptsCompls   = filtListWith mkExtCompl
+
+      stripLeading :: Char -> String -> String
+      stripLeading _ [] = []
+      stripLeading c (s:ss)
+        | s == c = ss
+        | otherwise = s:ss
+
+      result
+        | "import " `T.isPrefixOf` fullLine
+        = filtImportCompls
+        | "{-# language" `T.isPrefixOf` T.toLower fullLine
+        = filtOptsCompls languagesAndExts
+        | "{-# options_ghc" `T.isPrefixOf` T.toLower fullLine
+        = filtOptsCompls (map (T.pack . stripLeading '-') $ flagsForCompletion False)
+        | "{-# " `T.isPrefixOf` fullLine
+        = filtPragmaCompls (pragmaSuffix fullLine)
+        | otherwise
+        = filtModNameCompls ++ map (toggleSnippets caps withSnippets
+                                      . mkCompl . stripAutoGenerated) filtCompls
+  
+  return result
+
+-- The supported languages and extensions
+languagesAndExts :: [T.Text]
+languagesAndExts = map T.pack DynFlags.supportedLanguagesAndExtensions
+  
+-- ---------------------------------------------------------------------
+-- helper functions for pragmas
+-- ---------------------------------------------------------------------
+
+validPragmas :: [(T.Text, T.Text)]
+validPragmas =
```

Comment:
At least these require the `withSnippets` logic so let’s keep this for now.

Reply:
👍